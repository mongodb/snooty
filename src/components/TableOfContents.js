import React from 'react';
import PropTypes from 'prop-types';
import { css } from '@emotion/core';
import TOCNode from './TOCNode';
import { TOCContext } from './toc-context';

/**
 * Overall Table of Contents component, which manages open sections as children
 */
const TableOfContents = ({ activeSection, height, toctreeData: { children } }) => (
  <TOCContext.Provider value={{ activeSection }}>
    <ul
      className="current"
      css={css`
        height: calc(100% - ${height}px);
        overflow-y: auto;
        position: absolute;
        width: 100%;
      `}
    >
      {children.map((c) => {
        const key = c.slug || c.url;
        return <TOCNode node={c} key={key} />;
      })}
    </ul>
  </TOCContext.Provider>
);

TableOfContents.propTypes = {
  activeSection: PropTypes.string.isRequired,
  height: PropTypes.number.isRequired,
  toctreeData: PropTypes.shape({
    children: PropTypes.array.isRequired,
  }).isRequired,
};

export default TableOfContents;
