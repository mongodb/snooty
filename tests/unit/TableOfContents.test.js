import React from 'react';
import { mount } from 'enzyme';
import mockTocData from './data/Table-Of-Contents.test.json';
import TableOfContents from '../../src/components/TableOfContents';
/*
  Test render
  Test number of sections/children
  Test drawer specs
  Test options/styles
  Test links
*/

const mountedToc = (mockData, page) =>
  mount(<TableOfContents toctreeData={mockData} height={0} activeSection={page} />);

describe('Table of Contents testing', () => {
  describe('Table of Contents unit tests', () => {
    let testComponent;

    const remountTOC = (page) => {
      testComponent = mountedToc(mockTocData, page);
    };

    const updatePageLocation = (title, newLocation) => {
      remountTOC(newLocation);
    };

    beforeEach(() => {
      updatePageLocation('Ecosystem', '/');
      remountTOC();
    });

    it('renders correctly', () => {
      expect(testComponent).toMatchSnapshot();
    });

    it('TOC exists with proper number of sections and nodes per section', () => {
      // Child list of sections
      expect(testComponent.find('ul.current')).toHaveLength(1);
      // Child sections themselves
      expect(testComponent.find('ul li.toctree-l1')).toHaveLength(4);
      // Number in section
      expect(testComponent.find('.toctree-l2')).toHaveLength(0);
      updatePageLocation('Drivers', 'drivers');
      const numDriverNodes = mockTocData.children[0].children.length;
      expect(testComponent.find('.toctree-l2')).toHaveLength(numDriverNodes);
    });

    describe('TOC navigation should render and work as expected', () => {
      it('TOC slugs work as expected', () => {
        const testTOCLink = testComponent.find('ul li.toctree-l1 .reference').first();
        expect(testTOCLink.hasClass('internal')).toBe(true);
        expect(testTOCLink.prop('to')).toBe('drivers');
        expect(testComponent.find('.toctree-l2')).toHaveLength(0);
        updatePageLocation('Drivers', 'drivers');
        const numDriverNodes = mockTocData.children[0].children.length;
        expect(testComponent.find('.toctree-l2')).toHaveLength(numDriverNodes);
      });

      it('TOC external navigation (urls) should work as expected', () => {
        updatePageLocation('Tools', 'tools');
        const biConnectorLink = testComponent.find('.toctree-l2').first().find('a.reference');
        expect(biConnectorLink.hasClass('external')).toBe(true);
        expect(biConnectorLink.prop('href')).toBe('https://docs.mongodb.com/bi-connector/current/');
      });
    });

    describe('TOC options work as expected', () => {
      it('TOC supports a drawer option', () => {
        // Platforms section is a drawer, check clicking it does not redirect and opens those options
        expect(testComponent.find('.toctree-l2')).toHaveLength(0);
        const drawer = testComponent.find('.toctree-l1').last().find('.reference').first();
        drawer.simulate('click');
        const numUseCasesNodes = mockTocData.children[3].children.length;
        expect(testComponent.find('.toctree-l2')).toHaveLength(numUseCasesNodes);
        const otherDrawer = testComponent.find('.toctree-l1').at(2).find('.reference').first();
        otherDrawer.simulate('click');
        const numPlatformNodes = mockTocData.children[2].children.length;
        expect(testComponent.find('.toctree-l2')).toHaveLength(numUseCasesNodes + numPlatformNodes);
      });
    });
  });
});
