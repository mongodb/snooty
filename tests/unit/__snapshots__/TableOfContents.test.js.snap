// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table of Contents testing Table of Contents unit tests renders correctly 1`] = `
.emotion-0 {
  height: calc(100% - 0px);
  overflow-y: auto;
  position: absolute;
  width: 100%;
}

<TableOfContents
  height={0}
  toctreeData={
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "slug": "drivers/c",
              "title": "C Driver",
            },
            Object {
              "children": Array [],
              "slug": "drivers/cxx",
              "title": "C++ Driver",
            },
            Object {
              "children": Array [],
              "slug": "drivers/csharp",
              "title": "C# and .NET Driver",
            },
            Object {
              "children": Array [],
              "slug": "drivers/go",
              "title": "Go Driver",
            },
            Object {
              "children": Array [],
              "slug": "drivers/java",
              "title": "Java Driver",
            },
            Object {
              "children": Array [],
              "slug": "drivers/node",
              "title": "Node.js Driver",
            },
            Object {
              "children": Array [],
              "slug": "drivers/perl",
              "title": "Perl Driver",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "slug": "drivers/php-libraries",
                  "title": "PHP Libraries, Frameworks, and Tools",
                },
              ],
              "slug": "drivers/php",
              "title": "PHP Driver",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "slug": "drivers/pymongo",
                  "title": "PyMongo",
                },
                Object {
                  "children": Array [],
                  "slug": "drivers/motor",
                  "title": "Motor (Async Driver)",
                },
              ],
              "slug": "drivers/python",
              "title": "Python Drivers",
            },
            Object {
              "children": Array [],
              "title": "Ruby Driver",
              "url": "https://docs.mongodb.com/ruby-driver/current/",
            },
            Object {
              "children": Array [],
              "title": "Mongoid ODM",
              "url": "https://docs.mongodb.com/mongoid/current/",
            },
            Object {
              "children": Array [],
              "slug": "drivers/scala",
              "title": "Scala Driver",
            },
            Object {
              "children": Array [],
              "slug": "drivers/community-supported-drivers",
              "title": "Community Supported Drivers Reference",
            },
            Object {
              "children": Array [],
              "slug": "drivers/specs",
              "title": "MongoDB Drivers Specs",
            },
            Object {
              "children": Array [],
              "slug": "drivers/driver-compatibility-reference",
              "title": "Driver Compatibility",
            },
          ],
          "slug": "drivers",
          "title": "MongoDB Drivers and ODM",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "title": "MongoDB Connector for Business Intelligence",
              "url": "https://docs.mongodb.com/bi-connector/current/",
            },
            Object {
              "children": Array [],
              "title": "MongoDB ODBC Driver for BI Connector",
              "url": "https://docs.mongodb.com/bi-connector/current/reference/odbc-driver/",
            },
            Object {
              "children": Array [],
              "slug": "tools/http-interfaces",
              "title": "HTTP Interface",
            },
          ],
          "slug": "tools",
          "title": "MongoDB Integration and Tools",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "slug": "platforms/amazon-ec2",
              "title": "Amazon EC2",
            },
            Object {
              "children": Array [],
              "slug": "tutorial/backup-and-restore-mongodb-on-amazon-ec2",
              "title": "EC2 Backup and Restore",
            },
            Object {
              "children": Array [],
              "slug": "platforms/windows-azure",
              "title": "Microsoft Azure",
            },
            Object {
              "children": Array [],
              "slug": "platforms/windows",
              "title": "Windows Quick Links and Reference Center",
            },
          ],
          "options": Object {
            "drawer": true,
            "styles": Object {
              "code": "Platforms",
            },
          },
          "slug": "platforms",
          "title": "Platforms",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "slug": "use-cases/storing-log-data",
              "title": "Storing Log Data",
            },
            Object {
              "children": Array [],
              "slug": "use-cases/hierarchical-aggregation",
              "title": "Hierarchical Aggregation",
            },
            Object {
              "children": Array [],
              "slug": "use-cases/product-catalog",
              "title": "Product Catalog",
            },
            Object {
              "children": Array [],
              "slug": "use-cases/inventory-management",
              "title": "Inventory Management",
            },
            Object {
              "children": Array [],
              "slug": "use-cases/category-hierarchy",
              "title": "Category Hierarchy",
            },
            Object {
              "children": Array [],
              "slug": "use-cases/metadata-and-asset-management",
              "title": "Metadata and Asset Management",
            },
            Object {
              "children": Array [],
              "slug": "use-cases/storing-comments",
              "title": "Storing Comments",
            },
          ],
          "options": Object {
            "drawer": true,
            "styles": Object {
              "strong": "Use Cases",
            },
          },
          "slug": "use-cases",
          "title": "Use Cases",
        },
      ],
      "slug": "/",
      "title": "MongoDB Ecosystem",
    }
  }
>
  <ul
    className="current emotion-0"
  >
    <TOCNode
      key="drivers"
      level={1}
      node={
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "slug": "drivers/c",
              "title": "C Driver",
            },
            Object {
              "children": Array [],
              "slug": "drivers/cxx",
              "title": "C++ Driver",
            },
            Object {
              "children": Array [],
              "slug": "drivers/csharp",
              "title": "C# and .NET Driver",
            },
            Object {
              "children": Array [],
              "slug": "drivers/go",
              "title": "Go Driver",
            },
            Object {
              "children": Array [],
              "slug": "drivers/java",
              "title": "Java Driver",
            },
            Object {
              "children": Array [],
              "slug": "drivers/node",
              "title": "Node.js Driver",
            },
            Object {
              "children": Array [],
              "slug": "drivers/perl",
              "title": "Perl Driver",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "slug": "drivers/php-libraries",
                  "title": "PHP Libraries, Frameworks, and Tools",
                },
              ],
              "slug": "drivers/php",
              "title": "PHP Driver",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "slug": "drivers/pymongo",
                  "title": "PyMongo",
                },
                Object {
                  "children": Array [],
                  "slug": "drivers/motor",
                  "title": "Motor (Async Driver)",
                },
              ],
              "slug": "drivers/python",
              "title": "Python Drivers",
            },
            Object {
              "children": Array [],
              "title": "Ruby Driver",
              "url": "https://docs.mongodb.com/ruby-driver/current/",
            },
            Object {
              "children": Array [],
              "title": "Mongoid ODM",
              "url": "https://docs.mongodb.com/mongoid/current/",
            },
            Object {
              "children": Array [],
              "slug": "drivers/scala",
              "title": "Scala Driver",
            },
            Object {
              "children": Array [],
              "slug": "drivers/community-supported-drivers",
              "title": "Community Supported Drivers Reference",
            },
            Object {
              "children": Array [],
              "slug": "drivers/specs",
              "title": "MongoDB Drivers Specs",
            },
            Object {
              "children": Array [],
              "slug": "drivers/driver-compatibility-reference",
              "title": "Driver Compatibility",
            },
          ],
          "slug": "drivers",
          "title": "MongoDB Drivers and ODM",
        }
      }
    >
      <li
        className="toctree-l1  "
      >
        <NodeLink>
          <Link
            aria-expanded={false}
            className="reference  internal"
            to="drivers"
          >
            <mockConstructor
              aria-expanded={false}
              className="reference  internal"
              to="/drivers/"
            >
              <a
                aria-expanded={false}
                className="reference  internal"
                href="/drivers/"
              >
                MongoDB Drivers and ODM
              </a>
            </mockConstructor>
          </Link>
        </NodeLink>
      </li>
    </TOCNode>
    <TOCNode
      key="tools"
      level={1}
      node={
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "title": "MongoDB Connector for Business Intelligence",
              "url": "https://docs.mongodb.com/bi-connector/current/",
            },
            Object {
              "children": Array [],
              "title": "MongoDB ODBC Driver for BI Connector",
              "url": "https://docs.mongodb.com/bi-connector/current/reference/odbc-driver/",
            },
            Object {
              "children": Array [],
              "slug": "tools/http-interfaces",
              "title": "HTTP Interface",
            },
          ],
          "slug": "tools",
          "title": "MongoDB Integration and Tools",
        }
      }
    >
      <li
        className="toctree-l1  "
      >
        <NodeLink>
          <Link
            aria-expanded={false}
            className="reference  internal"
            to="tools"
          >
            <mockConstructor
              aria-expanded={false}
              className="reference  internal"
              to="/tools/"
            >
              <a
                aria-expanded={false}
                className="reference  internal"
                href="/tools/"
              >
                MongoDB Integration and Tools
              </a>
            </mockConstructor>
          </Link>
        </NodeLink>
      </li>
    </TOCNode>
    <TOCNode
      key="platforms"
      level={1}
      node={
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "slug": "platforms/amazon-ec2",
              "title": "Amazon EC2",
            },
            Object {
              "children": Array [],
              "slug": "tutorial/backup-and-restore-mongodb-on-amazon-ec2",
              "title": "EC2 Backup and Restore",
            },
            Object {
              "children": Array [],
              "slug": "platforms/windows-azure",
              "title": "Microsoft Azure",
            },
            Object {
              "children": Array [],
              "slug": "platforms/windows",
              "title": "Windows Quick Links and Reference Center",
            },
          ],
          "options": Object {
            "drawer": true,
            "styles": Object {
              "code": "Platforms",
            },
          },
          "slug": "platforms",
          "title": "Platforms",
        }
      }
    >
      <li
        className="toctree-l1  "
      >
        <NodeLink>
          <Link
            aria-expanded={false}
            className="reference  internal"
            href="platforms"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex="0"
          >
            <a
              aria-expanded={false}
              className="reference  internal"
              href="platforms"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex="0"
            >
              Platforms
            </a>
          </Link>
        </NodeLink>
      </li>
    </TOCNode>
    <TOCNode
      key="use-cases"
      level={1}
      node={
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "slug": "use-cases/storing-log-data",
              "title": "Storing Log Data",
            },
            Object {
              "children": Array [],
              "slug": "use-cases/hierarchical-aggregation",
              "title": "Hierarchical Aggregation",
            },
            Object {
              "children": Array [],
              "slug": "use-cases/product-catalog",
              "title": "Product Catalog",
            },
            Object {
              "children": Array [],
              "slug": "use-cases/inventory-management",
              "title": "Inventory Management",
            },
            Object {
              "children": Array [],
              "slug": "use-cases/category-hierarchy",
              "title": "Category Hierarchy",
            },
            Object {
              "children": Array [],
              "slug": "use-cases/metadata-and-asset-management",
              "title": "Metadata and Asset Management",
            },
            Object {
              "children": Array [],
              "slug": "use-cases/storing-comments",
              "title": "Storing Comments",
            },
          ],
          "options": Object {
            "drawer": true,
            "styles": Object {
              "strong": "Use Cases",
            },
          },
          "slug": "use-cases",
          "title": "Use Cases",
        }
      }
    >
      <li
        className="toctree-l1  "
      >
        <NodeLink>
          <Link
            aria-expanded={false}
            className="reference  internal"
            href="use-cases"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex="0"
          >
            <a
              aria-expanded={false}
              className="reference  internal"
              href="use-cases"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex="0"
            >
              Use Cases
            </a>
          </Link>
        </NodeLink>
      </li>
    </TOCNode>
  </ul>
</TableOfContents>
`;
